File: source/extensions/Extensions.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Reflection;
using System.Data;
using System.ComponentModel;
using System.Text.RegularExpressions;
using System.Collections.Concurrent;

namespace Sandbox.Utilities
{
    public static class Extensions
    {
        private static ConcurrentDictionary<Type, IList<PropertyInfo>> typeDictionary = new ConcurrentDictionary<Type, IList<PropertyInfo>>();

        public static IList<PropertyInfo> GetPropertiesForType<T>()
        {
            //variables
            var type = typeof(T);

            //add to dictionary
            typeDictionary.TryAdd(type, type.GetProperties().ToList());

            //return
            return typeDictionary[type];
        }

        public static T ToObject<T>(this DataRow row) where T : new()
        {
            // variables
            IList<PropertyInfo> properties = GetPropertiesForType<T>();

            // return
            return CreateItemFromRow<T>(row, properties);
        }

        public static IList<T> ToList<T>(this DataTable table) where T : new()
        {
            // variables
            IList<T> result = new List<T>();

            // foreach
            foreach (DataRow row in table.Rows)
            {
                result.Add(row.ToObject<T>());
            }

            // return
            return result;
        }

        private static T CreateItemFromRow<T>(DataRow row, IList<PropertyInfo> properties) where T : new()
        {
            // variables
            T item = new T();

            // foreach
            foreach (var property in properties)
            {
                // make sure a column exists in the table with this property name
                if (row.Table.Columns.Contains(property.Name))
                {
                    // get the value from the current data row
                    object value = row[property.Name];

                    // set property accordingly
                    if (value != null & value != DBNull.Value)
                    {
                        SetProperty<T>(item, property.Name, value);
                    }
                }
            }

            // return
            return item;
        }

        public static string GetProperty<T>(this T obj, string Property)
        {
            // reflection
            PropertyInfo propertyInfo = obj.GetType().GetProperty(Property, BindingFlags.Public | BindingFlags.Instance | BindingFlags.IgnoreCase);
            object property = null;

            // make sure property is valid
            if (propertyInfo != null)
            {
                property = propertyInfo.GetValue(obj, null);
            }

            // return value
            if (property != null)
            {
                return property.ToString();
            }
            else
            {
                return string.Empty;
            }
        }

        public static T SetProperty<T>(this T obj, string Property, object Value)
        {
            // reflection
            PropertyInfo prop = obj.GetType().GetProperty(Property, BindingFlags.Public | BindingFlags.Instance | BindingFlags.IgnoreCase);

            // trim strings
            if (Value.GetType() == typeof(string))
            {
                Value = Value.ToString().Trim();
            }

            // make sure property is valid
            if (prop != null && prop.CanWrite)
            {
                prop.SetValue(obj, Value, null);
            }

            // return
            return obj;
        }

        public static DataTable ToDataTable<T>(this IList<T> data)
        {
            // variables
            PropertyDescriptorCollection Properties = TypeDescriptor.GetProperties(typeof(T));
            object[] values = new object[Properties.Count];
            DataTable DT = new DataTable();

            // columns
            foreach (PropertyDescriptor PropertyInfo in Properties)
            {
                // data column
                DataColumn DataColumn = new DataColumn();

                // name
                DataColumn.ColumnName = PropertyInfo.Name;

                // data type
                if (PropertyInfo.PropertyType.Name.Contains("Nullable"))
                {
                    DataColumn.DataType = typeof(String);
                }
                else
                {
                    DataColumn.DataType = PropertyInfo.PropertyType;
                }

                // add to table
                DT.Columns.Add(DataColumn);
            }

            //for (int i = 0; i < Properties.Count; i++)
            //{
            //    PropertyDescriptor prop = Properties[i];
            //    DT.Columns.Add(prop.Name, prop.PropertyType);
            //}

            // rows
            foreach (T item in data)
            {
                for (int i = 0; i < values.Length; i++)
                {
                    values[i] = Properties[i].GetValue(item);
                }
                DT.Rows.Add(values);
            }

            // return
            return DT;
        }

        public static bool HasValue(this string Value)
        {
            return !Value.IsBlank();
        }

        public static bool IsBlank(this string Value)
        {
            bool ReturnValue = true;

            if (Value != null)
            {
                ReturnValue = Value.Trim().Length == 0;
            }

            return ReturnValue;
        }

        public static bool IsEqual(this string Value, string CompareValue)
        {
            bool ReturnValue = false;

            if (Value != null && CompareValue != null)
            {
                ReturnValue = string.Compare(Value.Trim(), CompareValue.Trim(), StringComparison.OrdinalIgnoreCase) == 0;
            }

            return ReturnValue;
        }

        public static bool IsEqual(this string Value, params string[] CompareValues)
        {
            if (Value != null && CompareValues != null)
            {
                foreach (string CompareValue in CompareValues)
                {
                    if (Value.IsEqual(CompareValue))
                    {
                        return true;
                    }
                }
            }

            return false;
        }

        public static bool ContainsValue(this string Value, string CompareValue)
        {
            bool ReturnValue = false;

            if (Value != null && CompareValue != null)
            {
                ReturnValue = Value.Trim().IndexOf(CompareValue.Trim(), StringComparison.OrdinalIgnoreCase) >= 0;
            }

            return ReturnValue;
        }

        public static bool ContainsValue(this string Value, params string[] CompareValues)
        {
            if (Value != null && CompareValues != null)
            {
                foreach (string CompareValue in CompareValues)
                {
                    if (Value.ContainsValue(CompareValue))
                    {
                        return true;
                    }
                }
            }

            return false;
        }

        public static bool StartsWithValue(this string Value, string CompareValue)
        {
            bool ReturnValue = false;

            if (Value != null && CompareValue != null)
            {
                ReturnValue = Value.Trim().StartsWith(CompareValue.Trim(), StringComparison.OrdinalIgnoreCase);
            }

            return ReturnValue;
        }

        public static bool StartsWithValue(this string Value, params string[] CompareValues)
        {
            if (Value != null && CompareValues != null)
            {
                foreach (string CompareValue in CompareValues)
                {
                    if (Value.StartsWithValue(CompareValue))
                    {
                        return true;
                    }
                }
            }

            return false;
        }

        public static string ReplaceValue(this string Value, string Pattern, string Replacement)
        {
            return Regex.Replace(Value, Regex.Escape(Pattern), Replacement, RegexOptions.IgnoreCase);
        }
    }
}

File: source/extensions/Extensions.vb
Imports System.Reflection
Imports System.Runtime.CompilerServices
Imports System.ComponentModel
Imports System.Text.RegularExpressions
Imports System.Collections.Concurrent

Public Module Extensions
    Private typeDictionary As New ConcurrentDictionary(Of Type, IList(Of PropertyInfo))()

    Public Function GetPropertiesForType(Of T)() As IList(Of PropertyInfo)
        'variables
        Dim type = GetType(T)

        'add to dictionary
        typeDictionary.TryAdd(type, type.GetProperties().ToList())

        'return
        Return typeDictionary(type)
    End Function

    <Extension()> _
    Public Function ToObject(Of T As New)(row As DataRow) As T
        'variables
        Dim properties As IList(Of PropertyInfo) = GetPropertiesForType(Of T)()

        'return
        Return row.CreateItemFromRow(Of T)(properties)
    End Function

    <Extension()> _
    Public Function ToList(Of T As New)(table As DataTable) As IList(Of T)
        'variables
        Dim result As IList(Of T) = New List(Of T)()

        'foreach
        For Each row As DataRow In table.Rows
            result.Add(row.ToObject(Of T)())
        Next

        'return
        Return result
    End Function

    <Extension()> _
    Private Function CreateItemFromRow(Of T As New)(row As DataRow, properties As IList(Of PropertyInfo)) As T
        'variables
        Dim item As New T()

        'foreach
        For Each p As PropertyInfo In properties
            'make sure a column exists in the table with this property name
            If row.Table.Columns.Contains(p.Name) Then
                'get the value from the current data row
                Dim value As Object = row(p.Name)

                'set property accordingly
                If value IsNot Nothing And value IsNot DBNull.Value Then
                    SetProperty(Of T)(item, p.Name, value)
                End If
            End If
        Next

        'return
        Return item
    End Function

    <Extension()> _
    Public Function GetProperty(Of T)(obj As T, propertyName As String) As String
        'reflection
        Dim propertyInfo As PropertyInfo = obj.GetType().GetProperty(propertyName, BindingFlags.Public Or BindingFlags.Instance Or BindingFlags.IgnoreCase)
        Dim propertyObj As Object = Nothing

        'make sure property is valid
        If propertyInfo IsNot Nothing Then
            propertyObj = propertyInfo.GetValue(obj, Nothing)
        End If

        'return value
        If propertyObj IsNot Nothing Then
            Return propertyObj.ToString()
        Else
            Return String.Empty
        End If
    End Function

    <Extension()> _
    Public Function SetProperty(Of T)(obj As T, propertyName As String, value As Object) As T
        'reflection
        Dim propertyInfo As PropertyInfo = obj.GetType().GetProperty(propertyName, BindingFlags.Public Or BindingFlags.Instance Or BindingFlags.IgnoreCase)

        'make sure property is valid
        If propertyInfo IsNot Nothing AndAlso propertyInfo.CanWrite Then
            propertyInfo.SetValue(obj, value, Nothing)
        End If

        'return
        Return obj
    End Function

    Public Function HasProperty(Of T As New)(propertyName) As Boolean
        'reflection
        Dim obj As T = New T()
        Dim propertyInfo As PropertyInfo = obj.GetType().GetProperty(propertyName, BindingFlags.Public Or BindingFlags.Instance Or BindingFlags.IgnoreCase)

        'see if property exists
        If propertyInfo IsNot Nothing Then
            Return True
        Else
            Return False
        End If
    End Function

    Private Function CreateDataTable(Of T)(Optional ByVal TableName As String = "") As DataTable
        'variables
        Dim Type As Type = GetType(T)
        Dim DT As New DataTable()
        Dim Properties As PropertyDescriptorCollection = TypeDescriptor.GetProperties(Type)

        'table name
        If TableName.HasValue Then
            DT.TableName = TableName
        Else
            DT.TableName = Type.Name
        End If

        'properties
        For Each p As PropertyDescriptor In Properties
            DT.Columns.Add(p.Name, If(Nullable.GetUnderlyingType(p.PropertyType), p.PropertyType))
        Next

        'return
        Return DT
    End Function

    <Extension()> _
    Public Function ToDataTable(Of T)(List As IList(Of T), Optional ByVal TableName As String = "") As DataTable
        'variables
        Dim DT As DataTable = CreateDataTable(Of T)(TableName)
        Dim Type As Type = GetType(T)
        Dim Properties As PropertyDescriptorCollection = TypeDescriptor.GetProperties(Type)

        'data
        For Each item As T In List
            'variables
            Dim row As DataRow = DT.NewRow()

            'property
            For Each p As PropertyDescriptor In Properties
                row(p.Name) = If(p.GetValue(item), DBNull.Value)
            Next

            'row
            DT.Rows.Add(row)
        Next

        'return
        Return DT
    End Function

    <Extension()> _
    Public Function HasValue(Value As String) As Boolean
        Return Not Value.IsBlank
    End Function

    <Extension()> _
    Public Function IsBlank(Value As String) As Boolean
        Dim ReturnValue As Boolean = True

        If Value IsNot Nothing Then
            ReturnValue = Value.Trim().Length = 0
        End If

        Return ReturnValue
    End Function

    <Extension()> _
    Public Function IsEqual(Value As String, CompareValue As String) As Boolean
        Dim ReturnValue As Boolean = False

        If Value IsNot Nothing AndAlso CompareValue IsNot Nothing Then
            ReturnValue = String.Compare(Value.Trim(), CompareValue.Trim(), StringComparison.OrdinalIgnoreCase) = 0
        End If

        Return ReturnValue
    End Function

    <Extension()> _
    Public Function IsEqual(ByVal Value As String, ByVal ParamArray CompareValues() As String) As Boolean
        If Value IsNot Nothing AndAlso CompareValues IsNot Nothing Then
            For Each CompareValue As String In CompareValues
                If Value.IsEqual(CompareValue) Then
                    Return True
                End If
            Next
        End If

        Return False
    End Function

    <Extension()> _
    Public Function ContainsValue(Value As String, CompareValue As String) As Boolean
        Dim ReturnValue As Boolean = False

        If Value IsNot Nothing AndAlso CompareValue IsNot Nothing Then
            ReturnValue = Value.Trim().IndexOf(CompareValue.Trim(), StringComparison.OrdinalIgnoreCase) >= 0
        End If

        Return ReturnValue
    End Function

    <Extension()> _
    Public Function StartsWithValue(Value As String, CompareValue As String) As Boolean
        Dim ReturnValue As Boolean = False

        If Value IsNot Nothing AndAlso CompareValue IsNot Nothing Then
            ReturnValue = Value.Trim().StartsWith(CompareValue.Trim(), StringComparison.OrdinalIgnoreCase)
        End If

        Return ReturnValue
    End Function

    <Extension()> _
    Public Function ReplaceValue(Value As String, Pattern As String, Replacement As String) As String
        Return Regex.Replace(Value, Regex.Escape(Pattern), Replacement, RegexOptions.IgnoreCase)
    End Function
End Module
